### DISCUSSION TITLE
Using Skyfield offline in a Juno Jupyter notebook

### DISCUSSION THREAD

**Discussion by alexanderflint** [skyfielders/python-skyfield#958] [Q&A]
I’m interested in using Skyfield to do celestial navigation calculations in fully offline mode, specifically using the mobile Juno app that enables pure Python Jupyter notebooks on iOS/iPadOS. It seems that the main barrier to this is that while Skyfield itself is pure Python, it has a required dependency in sgp4 that is not pure Python (sgp4 wraps a C++ library). Given that sgp4 is only needed for satellite ephemeris data, which isn’t needed for celestial navigation, is there a straightforward way to make Skyfield work without this dependency? Thanks for any thoughts!

**Comment by brandon-rhodes** [comment]
I just ran across the fact that `pip` has an option `--no-dependencies` that should let you install Skyfield without the `sgp4` library. If you want planetary ephemeris `.bsp` flies to work, you will need to `pip install jplephem`, but that library is pure Python and hopefully should not give you any trouble.

**Reply by alexanderflint** [reply]
Thank you! Juno does its own package management, so in the Juno Jupyter notebook, you can't use pip --no-dependencies. I can uninstall sgp4 in the package manager, but the installed version of Skyfield still expects sgp4. I'll touch base with Juno's developer about the potential for finer level control of package installations to enable the equivalent of --no-dependencies. Thanks again!

**Reply by brandon-rhodes** [reply]
The problem might be that you are trying to import symbols from the `skyfield.api` module, which assumes that all dependencies are installed. Instead, try reading through that file to find where it imports things from, and import the symbols from their original modules. As long as you don't import anything from `skyfield.sgp4lib`, there shouldn't be any problems with the `sgp4` package missing.

**Reply by alexanderflint** [reply]
You are, once again, absolutely correct. I've gone through the imports that were happening by way of the skyfield.api module and made them direct imports for each one. 

The only one that cause a bit of trouble is trying to import Loader and load_file from iokit, because this module includes a sgp4 import. As a proof-of-concept, I've just pasted iokit in as a Jupyter cell and commented out the sgp4 import, and this works:

https://colab.research.google.com/drive/1n2610a-KqcBdoSqFTlcp0dCz9Aj3zuL0?usp=sharing

I should be able to figure out how to save this cell as a .py file inside Juno's file structure, and then I should be able to import from this local file (to clean up the Jupyter notebook code).

**Comment by brandon-rhodes** [comment]
If you want to avoid the install and use of the `sgp4` package, take these two steps:

1. As discussed in my incomplete earlier answer, do the `pip install` of Skyfield with `--no-dependencies`, then install `jplephem` separately.
2. Then, use the following snippet of Python code before importing Skyfield.

```
import sys

module_type = type(sys)
module = module_type('skyfield.sgp4lib')
module.EarthSatellite = None
sys.modules['skyfield.sgp4lib'] = module
```

This will let you not only import `skyfield.iokit`, but even `import skyfield.api`, without needing the `sgp4` module, because it provides the one symbol that the rest of Skyfield needs to function. Try it out, and see if it solves your use case without your needing to cut-and-paste your own version of iokit!
